pkg/store/cache/caching_bucket.go

```diff
@@ -0,0 +1,389 @@
+}
+
+func DefaultCachingBucketConfig() CachingBucketConfig {
+	return CachingBucketConfig{
+		ChunkBlockSize:            16000, // Equal to max chunk size.
+		ChunkObjectSizeTTL:        24 * time.Hour,
+		ChunkBlockTTL:             24 * time.Hour,
+		MaxChunksGetRangeRequests: 1,
+	}
+}
+
+// Bucket implementation that provides some caching features, using knowledge about how Thanos accesses data.
+type CachingBucket struct {
+	bucket objstore.Bucket
+	cache  cache.Cache
+
+	config CachingBucketConfig
+
+	logger            log.Logger
+	cachedChunkBytes  prometheus.Counter
+	fetchedChunkBytes prometheus.Counter
+}
+
+func NewCachingBucket(b objstore.Bucket, c cache.Cache, chunks CachingBucketConfig, logger log.Logger, reg prometheus.Registerer) (*CachingBucket, error) {
+	if b == nil {
+		return nil, errors.New("bucket is nil")
+	}
+	if c == nil {
+		return nil, errors.New("cache is nil")
+	}
+
+	return &CachingBucket{
+		config: chunks,
+		bucket: b,
+		cache:  c,
+		logger: logger,
+
+		cachedChunkBytes: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "thanos_store_caching_bucket_cached_chunk_bytes_total",
+			Help: "Total number of chunk bytes used from cache",
+		}),
+		fetchedChunkBytes: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "thanos_store_caching_bucket_fetched_chunk_bytes_total",
+			Help: "Total number of chunk bytes fetched from storage",
+		}),
+	}, nil
+}
+
+func (cb *CachingBucket) Close() error {
+	return cb.bucket.Close()
+}
+
+func (cb *CachingBucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	return cb.bucket.Upload(ctx, name, r)
+}
+
+func (cb *CachingBucket) Delete(ctx context.Context, name string) error {
+	return cb.bucket.Delete(ctx, name)
+}
+
+func (cb *CachingBucket) Name() string {
+	return "caching: " + cb.bucket.Name()
+}
+
+func (cb *CachingBucket) WithExpectedErrs(expectedFunc objstore.IsOpFailureExpectedFunc) objstore.Bucket {
+	if ib, ok := cb.bucket.(objstore.InstrumentedBucket); ok {
+		// Make a copy, but replace bucket with instrumented one.
+		res := &CachingBucket{}
+		*res = *cb
+		res.bucket = ib.WithExpectedErrs(expectedFunc)
+		return res
+	}
+
+	// nothing else to do (?)
+	return cb
+}
+
+func (cb *CachingBucket) ReaderWithExpectedErrs(expectedFunc objstore.IsOpFailureExpectedFunc) objstore.BucketReader {
+	return cb.WithExpectedErrs(expectedFunc)
+}
+
+func (cb *CachingBucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	return cb.bucket.Iter(ctx, dir, f)
+}
+
+func (cb *CachingBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	return cb.bucket.Get(ctx, name)
+}
+
+var chunksMatcher = regexp.MustCompile(`^.*/chunks/\d+$`)
+
+func isTSDBChunkFile(name string) bool {
+	return chunksMatcher.MatchString(name)
+}
+
+func (cb *CachingBucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	if isTSDBChunkFile(name) && off >= 0 && length > 0 {
```

Comment:
> Why do we need to add the check `off >= 0 && length > 0`? Some thoughts:
> 
> 1. Can we consider `off < 0` == offset zero?
> 2. Given the first thing we do in `getRangeChunkFile()` is reading the object size, could we just set `length = object size` if `length < 0`?

Both conditions you describe are incorrect and I think hands-off approach is better. We let underlying bucket handle it.

Specifically I think that normalizing incorrect values is a bad idea, because then the behavior changes when used with and without caching bucket.